omrsysinfo_get_CPU_load() return a new error for insufficient data

thallium opened the pull request:
This is to determine if a call to `omrsysinfo_get_CPU_load` has only
one data point recorded.

A new error, `OMRPORT_ERROR_INSUFFICIENT_DATA`, is returned
in this scenario.

Also fixed a mistake in the description saying that the first two
invocations return `OMRPORT_ERROR_OPFAILED` and fix the time unit in the
description to `ms`.

Related: https://github.com/eclipse-openj9/openj9/pull/18451

thallium commented:
@tajila FYI

tajila commented on port/unix/omrsysinfo.c:
```diff
@@ -4642,7 +4642,7 @@ omrsysinfo_get_CPU_load(struct OMRPortLibrary *portLibrary, double *cpuLoad)
 		*oldestCPUTime = currentCPUTime;
 		*latestCPUTime = currentCPUTime;
-		return OMRPORT_ERROR_OPFAILED;
+		return OMRPORT_ERROR_INVALID;
```
Its safer to add a new error code for this, perhaps `OMRPORT_ERROR_INSUFFICIENT_DATA`

tajila commented:
@babsingh Please review these changes

babsingh commented:
jenkins build all

babsingh commented:
Currently, this PR only updates the `unix` impl. The `win32` impl also needs to be updated:
https://github.com/eclipse/omr/blob/8b19b8082de56b067c23bafd59dd767a58d8c02f/port/win32/omrsysinfo.c#L1271-L1275

babsingh commented:
Also, the function description needs to be updated to reflect the new error handling:
https://github.com/eclipse/omr/blob/8b19b8082de56b067c23bafd59dd767a58d8c02f/port/common/omrsysinfo.c#L643-L663

babsingh commented:
There is also the `zos` impl, which needs to be updated:
https://github.com/eclipse/omr/blob/8b19b8082de56b067c23bafd59dd767a58d8c02f/port/unix/omrsysinfo.c#L4671-L4678

thallium commented:
> Also, the function description needs to be updated to reflect the new error handling:
> 
> https://github.com/eclipse/omr/blob/8b19b8082de56b067c23bafd59dd767a58d8c02f/port/common/omrsysinfo.c#L643-L663

Why does it says "on the first two invocations"? Is that a typo?

babsingh commented:
> Why does it says "on the first two invocations"? Is that a typo?

I am unable to confirm by just reading the below design and implementation; it doesn't seem straightforward.
- Design/requirements: https://github.com/eclipse/omr/pull/3152#issuecomment-438347430
- Implementation (4 commits): https://github.com/eclipse/omr/pull/4702

@thallium Can you confirm by adding print statements for the return value in the existing test (see below)? I can run PR builds on all platforms once you have added debug code in the test. `on the first two invocations`: might be platform specific.

https://github.com/eclipse/omr/blob/8b19b8082de56b067c23bafd59dd767a58d8c02f/fvtest/porttest/si.cpp#L1689-L1716

babsingh commented:
@fjeremic For https://github.com/eclipse/omr/pull/7189#issuecomment-1823791887, do you remember why the second invocation of `omrsysinfo_get_CPU_load` also return a `OMRPORT_ERROR_OPFAILED`?

thallium commented:
@babsingh I've added the debug output, can you run the test?

babsingh commented:
jenkins build all(test:'-R porttest',env:'GTEST_FILTER=PortSysinfoTest.sysinfo_test_get_CPU_load',compile:'-j8')

If the `printf` output is not seen, then the `log` methods in the test suite will need to be used.

babsingh commented:
jenkins build all(test:'-R porttest',env:'GTEST_FILTER=PortSysinfoTest.sysinfo_test_get_CPU_load')

babsingh commented:
Windows compilation failed:

```
00:25:44      53>c:\omr\workspace\pullrequest-win_x86-64\build\fvtest\porttest\si.cpp(1705): error C2220: warning treated as error - no 'object' file generated [C:\omr\workspace\PullRequest-win_x86-64\Build\build\fvtest\porttest\omrporttest.vcxproj]
00:25:44      53>c:\omr\workspace\pullrequest-win_x86-64\build\fvtest\porttest\si.cpp(1705): warning C4244: '=': conversion from 'intptr_t' to 'int', possible loss of data [C:\omr\workspace\PullRequest-win_x86-64\Build\build\fvtest\porttest\omrporttest.vcxproj]
00:25:44      53>c:\omr\workspace\pullrequest-win_x86-64\build\fvtest\porttest\si.cpp(1707): warning C4244: '=': conversion from 'intptr_t' to 'int', possible loss of data [C:\omr\workspace\PullRequest-win_x86-64\Build\build\fvtest\porttest\omrporttest.vcxproj]
```

Test output:
```
00:43:49  15: Return value of first invocation: -21
00:43:49  15: Return value of second invocation: -1
```

I only checked a few build jobs. I consistently saw the above test output. @thallium Verify the test output in the PR builds as well. Fix the Windows compilation issue and verify the test output on Windows.

babsingh commented:
The test output for Windows is also the same:
```
00:43:49  15: Return value of first invocation: -21
00:43:49  15: Return value of second invocation: -1
```

The second `OMRPORT_ERROR_OPFAILED` is probably returned from the end of the function. @tajila I don't believe the current changes are sufficient to support https://github.com/eclipse-openj9/openj9/pull/18451 based upon the above behaviour.

thallium commented:
@babsingh Actually I think it's probably because the time between the two invocations are too short. In the description it says "10ns" but in the implementation what is check is `(currentCPUTime.timestamp - latestCPUTime->timestamp) >= 10000000` which is 10 milliseconds. I didn't see the two -1 on my local test:

```java
import java.lang.management.ManagementFactory;
import com.sun.management.OperatingSystemMXBean;

class CpuLoadTest{
  public static void main(String[] args) {
    while (true) {
        try
        {
            Thread.sleep(5000);
        }
        catch(InterruptedException ex)
        {
            Thread.currentThread().interrupt();
        }
        double cpuUtil = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getProcessCpuLoad();
        System.out.println(cpuUtil);
        cpuUtil = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getCpuLoad();
        System.out.println(cpuUtil);
    }
  }
}
```

but after I removed the sleep part, all the output became -1.

babsingh commented:
@thallium Good catch: the time units are wrong; there is insufficient data only in the first invocation; the error in the second invocation happens because we have waited for less than 10 ms. Can you update the units in the documentation and change `first two invocations` to `first invocation`?

babsingh commented on fvtest/porttest/si.cpp:
```diff
@@ -1690,19 +1690,17 @@ TEST(PortSysinfoTest, sysinfo_test_get_CPU_load)
-	 * as such the omrsysinfo_get_CPU_load could return a zero return code on the very first invocation within this
-	 * test.
+	 * that we cannot guarantee that no other test has called omrsysinfo_get_CPU_load up to this point. If some other 
+     * test did call these APIs then the internal buffers would have been populated and as such the omrsysinfo_get_CPU_load
+     * could return a zero return code on the very first invocation within this test.
```
indentation is off

babsingh commented on port/common/omrsysinfo.c:
```diff
@@ -651,10 +651,10 @@ omrsysinfo_get_CPU_utilization(struct OMRPortLibrary *portLibrary, struct J9Sysi
- *		- \arg OMRPORT_ERROR_OPFAILED if less than 10ns have passed since the second call to this API
- *		- negative portable error code on other failures
+ *  \arg 0 on success
+ *  \arg OMRPORT_ERROR_INSUFFICIENT_DATA on the first invocations of this API
```
```suggestion
 *  \arg OMRPORT_ERROR_INSUFFICIENT_DATA on the first invocation of this API
```

babsingh commented:
@thallium Also post a comment describing how you verified the correct behaviour for this function. One component is the test from https://github.com/eclipse/omr/pull/7189#issuecomment-1825890026, but also include the other steps that were used to verify the correct behaviour.

thallium commented:
> @babsingh Actually I think it's probably because the time between the two invocations are too short. In the description it says "10ns" but in the implementation what is check is `(currentCPUTime.timestamp - latestCPUTime->timestamp) >= 10000000` which is 10 milliseconds. I didn't see the two -1 on my local test:
> 
> ```java
> import java.lang.management.ManagementFactory;
> import com.sun.management.OperatingSystemMXBean;
> 
> class CpuLoadTest{
>   public static void main(String[] args) {
>     while (true) {
>         try
>         {
>             Thread.sleep(5000);
>         }
>         catch(InterruptedException ex)
>         {
>             Thread.currentThread().interrupt();
>         }
>         double cpuUtil = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getProcessCpuLoad();
>         System.out.println(cpuUtil);
>         cpuUtil = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getCpuLoad();
>         System.out.println(cpuUtil);
>     }
>   }
> }
> ```
> 
> but after I removed the sleep part, all the output became -1.

On top of the test, I added a print statement in `omrsysinfo_get_CPU_load()` to make sure no invocations happened outside the test. I only got negative return value on the first invocation.

babsingh commented:
Line endings check has failed:
```
###################################
fvtest/porttest/si.cpp:1695: trailing whitespace.
+	 * that we cannot guarantee that no other test has called omrsysinfo_get_CPU_load up to this point. If some other 
###################################
```

babsingh commented on fvtest/porttest/si.cpp:
```diff
@@ -1690,19 +1690,17 @@ TEST(PortSysinfoTest, sysinfo_test_get_CPU_load)
-	 * as such the omrsysinfo_get_CPU_load could return a zero return code on the very first invocation within this
-	 * test.
+	 * that we cannot guarantee that no other test has called omrsysinfo_get_CPU_load up to this point. If some other
+	 * test did call these APIs then the internal buffers would have been populated and as such the omrsysinfo_get_CPU_load
```
it is singular now since omrsysinfo_get_CPU_utlization has been removed from the previous line
```suggestion
	 * test calls this API then the internal buffers would have been populated and as such the omrsysinfo_get_CPU_load
```

babsingh commented:
jenkins build all

babsingh commented:
In your recent testing, `omrsysinfo_get_CPU_utilization` didn't fail so `omrsysinfo_get_CPU_load` returned `OMRPORT_ERROR_INSUFFICIENT_DATA` as the first error with the changes from this PR.

If `omrsysinfo_get_CPU_utilization` fails in the first invocation, then `omrsysinfo_get_CPU_load` will return the error code received from `omrsysinfo_get_CPU_utilization` instead of `OMRPORT_ERROR_INSUFFICIENT_DATA`.

There is also the case where `omrsysinfo_get_CPU_utilization` fails repeatedly before `oldestCPUTime->timestamp` is set.

In the above mentioned cases, the following statement in the function description no longer holds true: `return `OMRPORT_ERROR_INSUFFICIENT_DATA` on the first invocation of this API`.

@thallium How should we handle the above cases i.e. do we need to return `OMRPORT_ERROR_INSUFFICIENT_DATA` until `oldestCPUTime->timestamp` is set? If the above cases are encountered, will the changes proposed in https://github.com/eclipse-openj9/openj9/pull/18451 work as intended? 

https://github.com/eclipse/omr/blob/a18b161c95b93d19fe76fde378aa4bc515bec653/port/unix/omrsysinfo.c#L4635-L4646

thallium commented:
@babsingh Regarding the documentation, maybe we can change it to "on the first successful invocation". In the original issue, error handling was not discussed. I think we should return the error code from `omrsysinfo_get_CPU_utilization` but I could ask for people's opinion in the original issue.

babsingh commented:
> maybe we can change it to "on the first successful invocation"

This doesn't sound correct. No error should be thrown on the first **successful** invocation. 

> I think we should return the error code from omrsysinfo_get_CPU_utilization but I could ask for people's opinion in the original issue.

Sure, please ask in the original issue, and post an update in this issue once the expectations for error handling are confirmed.

tajila commented:
We need two data points to calculate CPU load. If we only have one, we should return insufficient data. If there is an error we should return an error code.

babsingh commented:
> If we only have one, we should return insufficient data. If there is an error we should return an error code.

If there is an error and insufficient data, which one takes precedence: error first, then insufficient data? Does OpenJ9 need to know that there is insufficient data when there is an error?

tajila commented:
> If there is an error and insufficient data, which one takes precedence: error first, then insufficient data? 

The precedent is which ever error is determined first is returned. What is an example where two errors can occur?

> Does OpenJ9 need to know that there is insufficient data when there is an error?

Openj9 just needs one of the error codes.

tajila commented:
> Intent of https://github.com/eclipse-openj9/openj9/pull/18451: Make get[Process|System]CpuLoad() return 0 on the first call.

I don't think this is fully accurate. The real intent is to return zero if we have one data point instead of -1 as the RI does. This is usually the case on the first call.

> can the RI still throw an error in the first call? 

Im not certain, ubt I would expect that to be the case. We can try to verify this. For example, if `/proc/stat` was not reachable then I would expect RI to fail as well and not return zero. 

If RI doesn't do this, then they are losing information, because returning 0 when there is a hard failure means there is no way to tell that there is a functional issue. In which case, I would still be okay with the behaviour being proposed here.

babsingh commented:
@thallium Please update the PR once the below tasks are completed.
- Update the documentation to reflect .... `The real intent is to return zero if we have one data point instead of -1 as the RI does.`
- Verify RI's behaviour if `/proc/stat` is not reachable.

thallium commented:
> Verify RI's behaviour if /proc/stat is not reachable.

RI returns -1 in this situation.

babsingh commented:
new line (return) is missing from the first line.

![image](https://github.com/eclipse/omr/assets/8634038/ee5a0a10-5371-40d4-a7bf-4e2b8e39ffa5)

babsingh commented:
still a space is missing `onlyone` -> `only one`. i have update the PR description for you.
![image](https://github.com/eclipse/omr/assets/8634038/fba7527a-9141-4c72-aa17-ac5a8a05b70f)

babsingh commented:
jenkins build aix

babsingh commented:
Only a known issue is observed:
- https://github.com/eclipse/omr/issues/6516
